### PR TITLE
Move `init_ddp_connection` to distributed utilities

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

It is duplicated in DDP & DDPSpawn plugins.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/distributed.py [hunk 1]
```diff
@@ -345,3 +346,26 @@ def register_ddp_comm_hook(
...
+    cluster_environment: ClusterEnvironment,
+    torch_distributed_backend: str,
+    global_rank: Optional[int] = None,
+    world_size: Optional[int] = None,
+) -> None:
+    global_rank = global_rank if global_rank is not None else cluster_environment.global_rank()
+    world_size = world_size if world_size is not None else cluster_environment.world_size()
+    os.environ["MASTER_ADDR"] = cluster_environment.master_address()
+    os.environ["MASTER_PORT"] = str(cluster_environment.master_port())
+    if torch.distributed.is_available() and not torch.distributed.is_initialized():
```
Comment (ananthsub) [inline pytorch_lightning/utilities/distributed.py:360]: should we raise a misconfiguration error if torch distributed isn't available?

File: pytorch_lightning/utilities/distributed.py [hunk 2]
```diff
@@ -345,3 +346,26 @@ def register_ddp_comm_hook(
...
 def tpu_distributed() -> bool:
     return _TPU_AVAILABLE and xm.xrt_world_size() > 1
+
+
+def init_ddp_connection(
+    cluster_environment: ClusterEnvironment,
+    torch_distributed_backend: str,
+    global_rank: Optional[int] = None,
+    world_size: Optional[int] = None,
+) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/utilities/distributed.py]: let's add a docstring for this as well?

File: pytorch_lightning/utilities/distributed.py [hunk 3]
```diff
@@ -345,3 +346,26 @@ def register_ddp_comm_hook(
 
 def tpu_distributed() -> bool:
     return _TPU_AVAILABLE and xm.xrt_world_size() > 1
+
+
+def init_ddp_connection(
+    cluster_environment: ClusterEnvironment,
```
Comment (ananthsub) [inline pytorch_lightning/utilities/distributed.py]: the precedence between cluster environment and global_rank/world_size might not be fully clear to callers

File: pytorch_lightning/utilities/distributed.py [hunk 4]
```diff
@@ -345,3 +346,26 @@ def register_ddp_comm_hook(
...
+    torch_distributed_backend: str,
+    global_rank: Optional[int] = None,
+    world_size: Optional[int] = None,
+) -> None:
+    global_rank = global_rank if global_rank is not None else cluster_environment.global_rank()
+    world_size = world_size if world_size is not None else cluster_environment.world_size()
+    os.environ["MASTER_ADDR"] = cluster_environment.master_address()
+    os.environ["MASTER_PORT"] = str(cluster_environment.master_port())
```
Comment (ananthsub) [inline pytorch_lightning/utilities/distributed.py:359]: i forget why we set this. is this relied on elsewhere in lightning? is it specific to the lightning cluster environment?

File: pytorch_lightning/utilities/distributed.py [hunk 5]
```diff
@@ -345,3 +346,26 @@ def register_ddp_comm_hook(
...
+    if torch.distributed.is_available() and not torch.distributed.is_initialized():
+        log.info(f"initializing ddp: GLOBAL_RANK: {global_rank}, MEMBER: {global_rank + 1}/{world_size}")
+        torch.distributed.init_process_group(torch_distributed_backend, rank=global_rank, world_size=world_size)
+
```
Comment (ananthsub) [inline pytorch_lightning/utilities/distributed.py:363]: init_process_group can take additional args, which we don't currently expose. should we make these settable with **kwargs? https://pytorch.org/docs/stable/_modules/torch/distributed/distributed_c10d.html#init_process_group

**Review by ananthsub — COMMENTED** [review]
fyi @four4fish as collective initialization could be part of https://github.com/PyTorchLightning/pytorch-lightning/issues/7534 

if the plugin initializes the global procss group, then we have higher confidence that the global process group can be destroyed in the teardown

File: pytorch_lightning/utilities/distributed.py (hunk 4 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/distributed.py:361]: Yes, it's specific to the Cluster Environments.

File: pytorch_lightning/utilities/distributed.py (hunk 5 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/distributed.py]: We could pass the kwargs from the `Plugins` constructor and then, to `init_process_group` by checking its signature and pass the matching params.

File: pytorch_lightning/utilities/distributed.py (hunk 3 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/distributed.py]: `global_rank/world_size` is Optional, hence the choice.

File: pytorch_lightning/utilities/distributed.py (hunk 2 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/distributed.py]: Added

File: pytorch_lightning/utilities/distributed.py (hunk 5 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/distributed.py]: Would be a follow-up PR to support it.

File: pytorch_lightning/utilities/distributed.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/distributed.py:373]: I think that's changing the behavior, we are not crashing it rn when we run it on an unsupported platform like Windows.

File: pytorch_lightning/utilities/distributed.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/utilities/distributed.py:372]: I believe this is required while creating the children with DDP ? @awaelchli  ?

File: pytorch_lightning/utilities/distributed.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/utilities/distributed.py:373]: If we decide to chance behaviour, let's do it in another PR.

**Review by tchaton — APPROVED** [review]
LGMT !

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

File: pytorch_lightning/utilities/distributed.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/distributed.py:372]: these are required for the `init_process_group` call below where we have `init_method='env://'`
the alternative is to do tcp initialization directly https://pytorch.org/docs/stable/distributed.html
Both ways are equivalent.

Unless something is broken here I would suggest not touching it ^^

**Review by awaelchli — APPROVED** [review]

**Comment by carmocca** [comment]
Hi @kaushikb11. Mind addressing the failing mypy job in a new PR?

**Pull request merged**